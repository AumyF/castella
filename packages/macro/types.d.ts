import React from "react";

declare const slotBland: unique symbol;

export type Slot<SlotName extends string> = {
  [slotBland]: SlotName;
};

/**
 * Converts given CSS string for use in Castella.
 */
export function css(
  arr: TemplateStringsArray,
  ...args: readonly unknown[]
): string;

export type HtmlTemplateItem<SlotName extends string> = string | Slot<SlotName>;

export type HtmlResult<SlotName extends string> = {
  shadowHtml: string;
  slots: readonly SlotName[];
};

/**
 * Converts given HTML string for use in Castella.
 */
export function html<SlotName extends string>(
  arr: TemplateStringsArray,
  ...args: readonly HtmlTemplateItem<SlotName>[]
): HtmlResult<Exclude<SlotName, "children" | "">>;

export type CastellaComponent<
  SlotName extends string
> = React.FunctionComponent<
  string extends SlotName
    ? {}
    : {
        [K in SlotName]?: React.ReactNode;
      }
>;

/**
 * Creates a slot.
 */
export function slot(): Slot<"">;
export function slot<SlotName extends string>(name: SlotName): Slot<SlotName>;

type ShadowAllowElementName =
  | "article"
  | "aside"
  | "blockquote"
  | "body"
  | "div"
  | "footer"
  | "h1"
  | "h2"
  | "h3"
  | "h4"
  | "h5"
  | "h6"
  | "header"
  | "main"
  | "nav"
  | "p"
  | "section"
  | "span";

type CastellaFunction = <SlotName extends string>(
  css: string,
  html: HtmlResult<SlotName>
) => CastellaComponent<SlotName>;

// export const castella: CastellaFunction &
//   Record<ShadowAllowElementName, CastellaFunction>;

/**
 * Creates a React component from given CSS and HTML.
 */
export const castella: CastellaFunction;

export const styled: {
  /**
   * Creates a React component with given CSS attached.
   */
  [Elm in keyof JSX.IntrinsicElements]: (
    arr: TemplateStringsArray,
    ...values: readonly string[]
  ) => React.FunctionComponent<JSX.IntrinsicElements[Elm]>;
};
