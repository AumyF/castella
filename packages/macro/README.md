# @castella/macro

Visit [the repository for Castella](https://github.com/uhyo/castella) for details.