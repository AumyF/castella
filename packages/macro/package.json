{
  "name": "@castella/macro",
  "version": "1.0.0",
  "description": "CSS in JS for React",
  "main": "lib/index.js",
  "types": "./types.d.ts",
  "scripts": {
    "build": "tsc",
    "watch": "tsc --watch",
    "test": "jest",
    "test:update": "jest -u",
    "prepublishOnly": "npm run build && npm run test"
  },
  "keywords": [
    "babel-plugin-macros"
  ],
  "author": "uhyo <uhyo@uhy.ooo>",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/uhyo/castella.git"
  },
  "files": [
    "lib/**/*",
    "src/**/*",
    "types.d.ts"
  ],
  "dependencies": {
    "babel-plugin-macros": "^2.8.0",
    "stylis": "^4.0.3"
  },
  "devDependencies": {
    "@babel/core": "^7.12.3",
    "@babel/types": "^7.12.1",
    "@types/babel-plugin-macros": "^2.8.3",
    "@types/babel-types": "^7.0.9",
    "@types/node": "^14.11.10",
    "@types/react": "^16.9.53",
    "@types/stylis": "^4.0.0",
    "babel-plugin-tester": "^10.0.0",
    "jest": "^26.5.3",
    "typescript": "^4.0.3"
  },
  "peerDependencies": {
    "@types/react": "^16.9.53"
  }
}