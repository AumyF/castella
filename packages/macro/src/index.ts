import { createMacro } from "babel-plugin-macros";

// `source` is not in @types/babel-plugin-macros :(
// @ts-expect-error
export = createMacro(({ references, state, babel, source }) => {
  const cssFuncReferences = [...(references.css || [])];
  const htmlFuncReferences = [...(references.html || [])];
  const wc = [...(references.wc || [])];
});
