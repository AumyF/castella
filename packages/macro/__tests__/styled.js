const pluginTester = require("babel-plugin-tester").default;
const plugin = require("babel-plugin-macros");

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename },
  tests: {
    "no hole": `
      import { styled } from '../../macro'

      const input = styled.input\`
        background-color: yellow;
      \`;
    `,
    "some hole": `
      import { styled } from '../../macro'

      const color = "yellow";
      const input = styled["input"]\`
        background-color: \${color};
      \`;
    `,
    "computed name": `
      import { styled } from '../../macro'

      const expr = "span";
      const あいうComponent = styled[expr]\`
        font-weight: bold;
      \`;
    `,
    "nested": `
      import { styled } from '../../macro'

      const あいう = styled.div\`
        display: flex;
        & > * {
          flex: 100px 0 0;
        }
      \`;
    `
  },
});


