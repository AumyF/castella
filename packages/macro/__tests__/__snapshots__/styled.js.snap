// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin computed name: computed name 1`] = `

import { styled } from '../../macro'

const expr = "span";
const あいうComponent = styled[expr]\`
  font-weight: bold;
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import { styledComponent as _styledComponent } from "@castella/runtime";
const expr = "span";

const あいうComponent = _styledComponent({
  css: \`.castella-あいうComponent-203ffa{font-weight:bold;}\`,
  className: "castella-\\u3042\\u3044\\u3046Component-203ffa",
  element: expr,
});


`;

exports[`unknown plugin nested: nested 1`] = `

import { styled } from '../../macro'

const あいう = styled.div\`
  display: flex;
  & > * {
    flex: 100px 0 0;
  }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import { styledComponent as _styledComponent } from "@castella/runtime";

const あいう = _styledComponent({
  css: \`.castella-あいう-203ffa{display:flex;}.castella-あいう-203ffa>*{flex:100px 0 0;}\`,
  className: "castella-\\u3042\\u3044\\u3046-203ffa",
  element: "div",
});


`;

exports[`unknown plugin no hole: no hole 1`] = `

import { styled } from '../../macro'

const input = styled.input\`
  background-color: yellow;
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import { styledComponent as _styledComponent } from "@castella/runtime";

const input = _styledComponent({
  css: \`.castella-input-203ffa{background-color:yellow;}\`,
  className: "castella-input-203ffa",
  element: "input",
});


`;

exports[`unknown plugin some hole: some hole 1`] = `

import { styled } from '../../macro'

const color = "yellow";
const input = styled["input"]\`
  background-color: \${color};
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import { styledComponent as _styledComponent } from "@castella/runtime";
const color = "yellow";

const input = _styledComponent({
  css: \`.castella-input-203ffa{background-color:\${color};}\`,
  className: "castella-input-203ffa",
  element: "input",
});


`;
